project4/proto/pub_sub_messages.proto

```diff
@@ -0,0 +1,140 @@
+}
+
+/// entrance request sent by the entering name server to the bootstrap name server.
+message EntranceRequest {
+  /// the id of the entering name server
+  uint32 id = 1;
+}
+
+/// entrance information to be rippled through the name servers.
+message EntranceInformation {
+  /// the id of the entering name server
+  uint32 id = 1;
+  /// the id of the successor of the entering name server
+  uint32 successor_id = 2;
+  /// the machine name of the entering name server's successor
+  string successor_ip = 3;
+  /// the port number the entering name server's successor is listening on
+  uint32 successor_port = 4;
+  /// the id of the predecessor of the entering name server
+  uint32 predecessor_id = 5;
+  /// the machine name of the entering name server's predecessor
+  string predecessor_ip = 6;
+  /// the port number the entering name server's predecessor is listening on
+  uint32 predecessor_port = 7;
+}
+
+/// request to update predecessor to be sent by the entering name server to it's to-be successor.
+message UpdatePredecessorRequest {
+  /// the id of the name server's new predecessor
+  uint32 id = 1;
+  /// the ip of the name server's new predecessor
+  string ip = 2;
+  /// the port the name server's new predecessor is listening on
+  uint32 port = 3;
+}
+
+/// exit information sent by the exiting name server to it's successor
+message ExitInformation {
+  /// the lower bounds of the exiting name server's key range.
+  uint32 lower_bounds = 1;
+  /// the upper bounds of the exiting name server's key range.
+  uint32 upper_bounds = 2;
+  /// the currently stored keys.
+  repeated uint32 keys = 3;
+  /// the values associated with the currently stored keys.
+  repeated string values = 4;
+  /// the id of the successor's new predecessor
+  uint32 id = 5;
+  /// the ip of the successor's new predecessor
+  string ip = 6;
+  /// the port the successor's new predecessor listens on.
+  uint32 port = 7;
+}
+
+/// response for updating predecessor.
+/// sent by the entering name server's successor to the entering name server.
+message UpdatePredecessorResponse {
+  /// the lower bounds of the entering name server's key range.
+  uint32 lower_bounds = 1;
+  /// the upper bounds of the entering name server's key range.
+  uint32 upper_bounds = 2;
+  /// the currently stored keys.
+  repeated uint32 keys = 3;
+  /// the values associated with the currently stored keys.
+  repeated string values = 4;
+
+}
+/// request to update successor info.
+/// sent by an entering name server to its to-be predecessor.
+/// or sent by an exiting name server to it's predecessor.
+message UpdateSuccessorRequest {
+  /// the id of the predecessor's new successor.
+  uint32 id = 1;
+  /// the ip of the predecessor's new successor.
+  string ip = 2;
+  /// the port the predecessor's new successor is listening on.
+  uint32 port = 3;
+}
+/// encapsulates all messages that can be received by the bootstrap name server
+message BootstrapMessage {
```

Comment:
Perhaps a tidier way to do this would be to have one of the options in the `oneof` be a `NameServerMessage`, and then everything else in the `oneof` are messages unique to the bootstrap node.